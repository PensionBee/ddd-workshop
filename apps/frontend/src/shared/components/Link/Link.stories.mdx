import { Meta, Story, Canvas, ArgsTable } from "@storybook/addon-docs";
import { Link } from "./Link";

<Meta
  title="Shared/Link/Docs"
  component={Link}
  parameters={{
    viewMode: "docs",
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
/>

# Link

## Props

<ArgsTable of={Link} />

## Usage & Customisation

### Default Link

<Canvas>
  <Story id="shared-link--default-link" />
</Canvas>

### Nested Link

<Canvas>
  <Story id="shared-link--nested-link" />
</Canvas>
