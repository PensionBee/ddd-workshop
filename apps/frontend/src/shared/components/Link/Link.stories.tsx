import type { ComponentMeta } from "@storybook/react";
import { Link } from "./Link";
import { BodySmallL } from "../Typography/Typography";

export default {
  title: "Shared/Link",
  component: Link,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof Link>;

export const DefaultLink = () => (
  <BodySmallL>
    <Link href="https://google.com" target="_blank">
      A link that can be clicked on.
    </Link>
  </BodySmallL>
);

export const NestedLink = () => (
  <BodySmallL>
    Lorem ipsum dolor sit amet consectetur adipisicing elit.{" "}
    <Link href="https://google.com" target="_blank">
      A nested link should be underlined.
    </Link>{" "}
    Necessitatibus maxime architecto mollitia maiores! Aliquam, sint facere!
  </BodySmallL>
);
