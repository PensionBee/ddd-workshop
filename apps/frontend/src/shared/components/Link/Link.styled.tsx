import { theme } from "@shared/styling/theme";
import styled from "styled-components";

export type StyledLinkProps = {
  $color: "link" | "black";
  $noUnderline: boolean;
  $nowrap: boolean;
};

export const StyledLink = styled.a<StyledLinkProps>`
  cursor: pointer;
  transition: color 200ms ease;
  color: ${({ $color }) => theme.colors[$color]};
  white-space: ${({ $nowrap }) => ($nowrap ? "nowrap" : "normal")};
  text-decoration: ${({ $noUnderline }) =>
    $noUnderline ? "none" : "underline"};
  text-underline-position: under;
  text-underline-offset: 5%;
  &:hover,
  &:focus-visible {
    color: ${({ $color }) => {
      if ($color === "black") return theme.colors.grey1;
      return theme.colors.linkActive;
    }};
    text-decoration: none;
  }
  &:focus-visible {
    outline: 3px solid ${() => theme.colors.blue100};
    border-radius: 4px;
  }
`;
