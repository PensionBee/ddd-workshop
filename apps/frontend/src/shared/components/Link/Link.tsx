import type { StyledLinkProps } from "./Link.styled";
import { StyledLink } from "./Link.styled";

export type LinkProps = {
  /** React children */
  children: React.ReactNode;
  /** Link color */
  color?: StyledLinkProps["$color"];
  /** Link href */
  href?: React.HTMLProps<HTMLAnchorElement>["href"];
  /** Defines if the link should be underlined - use this if the link is part of a larger body of copy */
  noUnderline?: StyledLinkProps["$noUnderline"];
  /** If true, prevents portion of text from wrapping, forcing the entire content onto a new line */
  nowrap?: StyledLinkProps["$nowrap"];
  /** Link target */
  target?: "_blank" | "_self" | "_parent" | "_top" | "framename";
};

export const Link: React.FC<LinkProps> = ({
  children,
  color = "link",
  href,
  noUnderline = false,
  nowrap = false,
  target = "_self",
  ...props
}) => {
  const styledProps = {
    $color: color,
    $nowrap: nowrap,
    $noUnderline: noUnderline,
  };

  return (
    <StyledLink
      data-rc="Link"
      href={href}
      target={target}
      {...styledProps}
      {...props}
    >
      {children}
    </StyledLink>
  );
};
