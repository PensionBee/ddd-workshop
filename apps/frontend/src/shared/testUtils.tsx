import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
