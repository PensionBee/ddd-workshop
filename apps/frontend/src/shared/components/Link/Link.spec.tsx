import { render } from "@shared/testUtils";
import { Link } from "./Link";

describe("Link", () => {
  const linkText = "Link text";
  const linkUrl = "/test";
  describe("Default", () => {
    it("should render correctly", () => {
      const { getByText, asFragment } = render(
        <Link href={linkUrl}>{linkText}</Link>
      );

      expect(getByText(linkText)).toBeVisible();
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
