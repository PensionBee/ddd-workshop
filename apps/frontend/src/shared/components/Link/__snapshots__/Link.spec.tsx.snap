// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Default should render correctly 1`] = `
<DocumentFragment>
  <a
    class="sc-bcXHqe gDMKLX"
    data-rc="Link"
    href="/test"
    target="_self"
  >
    Link text
  </a>
</DocumentFragment>
`;
